Dependencies/UrlLib/Source/Unix/UrlRequest.cpp

```diff
@@ -174,7 +174,58 @@ namespace UrlLib
             if (curl)
             {
                 data.clear();
-                curl_easy_setopt(curl, CURLOPT_URL, m_url.data());
+
+                CURLUcode rc;
+                CURLU *url = curl_url();
+                rc = curl_url_set(url, CURLUPART_URL, m_url.c_str(), 0);
+                if (rc != CURLUE_OK)
+                {
+                    throw std::runtime_error("CURL: Unable to build URL.");
+                }
+
+                char *scheme;
+                rc = curl_url_get(url, CURLUPART_SCHEME, &scheme, 0);
+                if (rc != CURLUE_OK)
+                {
+                    throw std::runtime_error("CURL: Unable to get URL scheme.");
+                }
+
+                if (strcmp(scheme, "file"))
+                {
+                    char *path;
+                    rc = curl_url_get(url, CURLUPART_PATH, &path, 0);
+                    if (rc != CURLUE_OK)
+                    {
+                        throw std::runtime_error("CURL: Unable to get URL path.");
+                    }
+
+                    char* pathEscaped = curl_easy_escape(curl, path, 0);
+                    rc = curl_url_set(url, CURLUPART_PATH, pathEscaped, 0);
+                    if (rc != CURLUE_OK)
+                    {
+                        throw std::runtime_error("CURL: Unable to set URL path.");
+                    }
+                    curl_free(pathEscaped);
+                    curl_free(path);
+
+                    char *urlEscaped;
+                    rc = curl_url_get(url, CURLUPART_URL, &urlEscaped, 0);
+                    if (rc != CURLUE_OK)
+                    {
+                        throw std::runtime_error("CURL: Unable to get URL string.");
+                    }
+
+                    curl_easy_setopt(curl, CURLOPT_URL, urlEscaped);
+                    curl_free(urlEscaped);
+                }
+                else
+                {
+                    curl_easy_setopt(curl, CURLOPT_URL, m_url.c_str());
+                }
+
+                curl_free(scheme);
+                curl_url_cleanup(url);
```

Comment:
Because of the throws earlier, it's probably good to put these clean up statements in gsl::finally so that they will get called even if an exception is thrown.

Reply:
Applies to all the `curl_free` and other clean up functions.